### PR TITLE
Added support for external Hive Metastores including AWS Glue

### LINKED ISSUE #337: Migrate `EXTERNAL` tables from AWS Glue or other external HMS
See https://github.com/databrickslabs/ucx/issues/317

# Definition of done:

- [x] full end-user documentation - README.md + revise [this document](https://github.com/databrickslabs/ucx/blob/main/docs/external_hms_glue.md) - #1458
- [x] integration test for AWS Glue - #1408, #1368
- [x] integration test for MSSQL - #1408, #1368
- [x] integration test for assessment -> table-migration for external tables #1408, #1460

### PR DESCRIPTION
**Pull request by FastLee**
Closes #337

### CODE REVIEW

File: src/databricks/labs/ucx/config.py [hunk 1]
```diff
@@ -207,9 +207,10 @@ class WorkspaceConfig(_Config["WorkspaceConfig"]):
     log_level: str | None = "INFO"
     database_to_catalog_mapping: dict[str, str] = None
     default_catalog: str = "ucx_default"
-
     # Starting path for notebooks and directories crawler
     workspace_start_path: str = "/"
+    instance_profile: str = None
```
Comment (nfx) [inline src/databricks/labs/ucx/config.py:212]: This config name is too generic

File: src/databricks/labs/ucx/install.py [hunk 2]
```diff
@@ -233,12 +234,36 @@ def warehouse_type(_):
...
+
+        instance_profile = None
+        spark_conf_dict = {}
+        if (
+            self._prompts
+            and self._question("Do you need to configure an external Hive Metastore (Glue)", default="no") == "yes"
+        ):
+            logger.info("Setting up an external metastore")
+            instance_profiles = self._instance_profiles()
+            instance_profile = ""
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:246]: Is it possible to detect that? Eg seeing the special spark confs in one of the clusters and using it as a template?

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/install.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:246]: It's possible. Maybe from the warehouse setting.

Comment (FastLee) [reply src/databricks/labs/ucx/install.py:246]: Let's start with the manual option and then extend to automate it.

File: src/databricks/labs/ucx/install.py [hunk 3]
```diff
@@ -233,12 +234,36 @@ def warehouse_type(_):
             groups_config_args["selected"] = [x.strip() for x in selected_groups.split(",")]
         else:
             groups_config_args["auto"] = True
+
+        instance_profile = None
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:238]: i think the best UX for this is a drop-down for AWS Glue-enabled clusters and keep the `aws_glue_cluster_id` as the private field of the installer. and then create special job tasks in the assessment to crawl for tables and tacls from AWS Glue metastore.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/install.py (hunk 3 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:238]: We don't need a special job. The existing one works with Glue (I confirmed it).

File: src/databricks/labs/ucx/install.py [hunk 4]
```diff
@@ -486,6 +524,17 @@ def _job_settings(self, step_name: str, dbfs_path: str):
             "tasks": [self._job_task(task, dbfs_path) for task in tasks],
         }
 
+    @staticmethod
+    def _apply_cluster_overrides(settings: dict[str, any], overrides: dict[str, str]) -> dict:
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:528]: did you change it? why does it appear in the diff?

File: src/databricks/labs/ucx/install.py [hunk 5]
```diff
@@ -285,12 +287,48 @@ def warehouse_type(_):
             groups_config_args["selected"] = [x.strip() for x in selected_groups.split(",")]
         else:
             groups_config_args["auto"] = True
+
+        # Checking for external HMS
+        instance_profile = None
+        spark_conf_dict = {}
+        if self._prompts:
+            ext_hms_confs = list(self._get_ext_hms_confs())
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:295]:
```suggestion
            policies_with_external_hms = list(self._get_ext_hms_confs())
```

File: src/databricks/labs/ucx/install.py [hunk 6]
```diff
@@ -285,12 +287,48 @@ def warehouse_type(_):
             groups_config_args["selected"] = [x.strip() for x in selected_groups.split(",")]
         else:
             groups_config_args["auto"] = True
+
+        # Checking for external HMS
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:291]: can you put this additional logic into separate private method, that returns spark conf and instance profile?

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -697,6 +755,27 @@ def _deployed_steps(self):
...
 
+    def _instance_profiles(self):
+        return {"No Instance Profile": None} | {
+            profile.instance_profile_arn: profile.instance_profile_arn for profile in self._ws.instance_profiles.list()
+        }
+
+    def _instance_profiles(self):
+        return {"No Instance Profile": None} | {
+            profile.instance_profile_arn: profile.instance_profile_arn for profile in self._ws.instance_profiles.list()
+        }
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:766]: Bad merge. Fixed

File: src/databricks/labs/ucx/install.py [hunk 8]
```diff
@@ -697,6 +755,27 @@ def _deployed_steps(self):
...
+
+    def _get_ext_hms_confs(self):
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:768]: Agreed

File: src/databricks/labs/ucx/install.py [hunk 9]
```diff
@@ -697,6 +755,27 @@ def _deployed_steps(self):
...
+        for conf in self._ws.cluster_policies.list():
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:769]: Done

File: src/databricks/labs/ucx/install.py (hunk 4 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:528]: I did.

File: src/databricks/labs/ucx/install.py (hunk 5 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:295]: Done

File: src/databricks/labs/ucx/install.py (hunk 6 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:291]: Sure

**Review by nfx — APPROVED** [review]
Make sure integration tests pass and then you can merge

**Pull request merged**